Makefile

```diff
@@ -70,16 +92,37 @@ deps: go.mod go.sum
 	go mod verify
 
 .PHONY: format
-format: $(GOLANGCI_LINT)
-	$(GOLANGCI_LINT) run --fix --enable-all -c .golangci.yml
+format: ## Formats Go code.
+format: $(GOIMPORTS) $(GOLANGCI_LINT)
+	@echo ">> formatting code"
+	@gofmt -s -w $(FILES_TO_FMT)
+	@$(GOIMPORTS) -w $(FILES_TO_FMT)
+
+.PHONY: check-git
+check-git:
+ifneq ($(GIT),)
+	@test -x $(GIT) || (echo >&2 "No git executable binary found at $(GIT)."; exit 1)
+else
+	@echo >&2 "No git binary found."; exit 1
+endif
 
 .PHONY: shellcheck
 shellcheck: $(SHELLCHECK)
 	$(SHELLCHECK) $(shell find . -type f -name "*.sh" -not -path "*vendor*" -not -path "*tmp*")
 
 .PHONY: lint
-lint: $(GOLANGCI_LINT) deps shellcheck jsonnet-fmt
-	$(GOLANGCI_LINT) run -v --enable-all -c .golangci.yml
+lint: ## Runs various static analysis against our code.
+lint: $(FAILLINT) $(GOLANGCI_LINT) $(MISSPELL) build docs check-git deps shellcheck jsonnet-fmt format
```

Comment:
@saswatamcode Could we not make `lint` depend on `check-git`? This basically forces people to commit to be able to run the `lint` target and it creates annoying ergonomics, i.e. forcing people to use git flows they might not be used to or simply coming to the Makefile and removing the `check-git` code (which I literally do 100% of the time).

Reply:
Also seems weird to me that we run all these linters but still have to run the actual `build` and then after building we run `deps and then `format`.... it seems too much and completely out of order.

Reply:
I agree about the ordering, it seems like it isn't accurate as I took it from obsctl, I will update this to be specific.

About `check-git`, that only checks if you have git installed. 
I think you meant `require_clean_work_tree`, which is pretty standard and used across many of our upstream projects like Thanos. The flow is made for CI too and is ideally something like generate+format+build -> check if that changes anything in tree -> lint -> check if linters change anything in tree. If it fails at any step, the next ones aren't run and you get notified by CI.

It doesn't force anything though, and it is just convenient to commit code and then lint to detect issues and fix them in one go? Or just push and fix if and when CI fails?
And we squash merge as it is, so extraneous commits are not problematic.

Wdyt? 🙂

Reply:
I meant the `require_clean_work_tree` indeed. If CI needs a clean work tree check after linting I believe it should call a separate `make check-work-tree` target after linting. Why? Because we're mixing things. I absolutely do not need a clean work tree to run linters, it's an arbitrary requirement that creates an obstacle to linting the project. 

Some might even say linters aren't supposed to format code, this is responsibility of formatters, hinting that they should be separate independent Makefile targets. But this is not the point now. 

> It doesn't force anything though, and it is just convenient to commit code and then lint to detect issues and fix them in one go? Or just push and fix if and when CI fails?

It's forcing people to commit before running linters, **unless** they remove it in their local copies. Some people might enjoy working with committing all the time and squashing, ammending, doing fixups, push `--force-with-lease`, etc... but others might not. For example, I prefer to code, lint the project with a dirty work tree, fix the issues, then commit. This is why you often see commits where I accidentally comment out the `require_clean_work_tree` check.